### PR TITLE
wasi-common support for tokio, & wiggle support for async methods containing sync code

### PR DESCRIPTION
**Pull request by pchickey**
This PR makes it possible for `wasi-common` to work on with async executor like `tokio`, without taking a dependency on any sort of asynchronous runtime. Existing users of `wasi-common` will not need to make any changes.

It introduces the new crate `wasi-tokio`, which is a sibling crate to `wasi-cap-std-sync`, and also depending on `wasi-cap-std-sync`, since much of the implementation has been reused.

`wasmtime-wiggle` now supports running async methods via a dummy executor, in order to support running synchronous wasi-common implementations. To configure this mode in `wasmtime-wiggle`, use the new `block_on` configuration option in place of `async` as introduced in #2701. This mode executes the async methods in a "dummy executor" - a mocked `std::task::Waker` which will only poll a future correctly if it immediately returns `Poll::Ready`.

`wasi-common` now uses an `async_trait` to make each `WasiFile`, `WasiDir`, and `WasiSched` trait method `async`. This means that any operation on files, directories, or the scheduler may return a pending future depending on the impl of those traits.

`wasi-cap-std-sync` stays the same, modulo the code golf required to add `async` in front of some fns. All of these impls are synchronous - the futures returned by these will always be `Poll::Ready`. Therefore, it is safe to use with the dummy executor.

The `WasiFile` trait has two additional methods, `async fn readable(&mut self) -> Result<(), Error>` and a corresponding `writable`, which have semantics corresponding to a wasi `Subscription::FdRead` and `FdWrite` - the futures are ready when the file is ready to read or write. These methods are intended to be used by an impl of `WasiSched::poll_oneoff`.

`wasi-tokio` is implemented mostly in terms of `wasi-cap-std-sync`. cap-std provides a robust implementation of filesystem sandboxing, which we still need to rely on. All file and directory operations delegated to the `wasi-cap-std-sync` impl are wrapped in `tokio::task::block_in_place`, which tells the tokio runtime to execute that code on a thread that may block. This is how tokio's own `File` type treats all of its IO operations as well. The departure of `wasi-tokio` from wasi-cap-std-sync is in the impl of the `WasiFile::{readable, writable}` methods, which use `tokio::io::unix::AsyncFd` when available, and the `WasiSched::poll_oneoff` method is implemented using those futures and `tokio::time::timeout`. 

Unfortunately, tokio's AsyncFd doesn't do very much in this context on Linux (`epoll(2)` doesn't operate on regular files), and isn't available at all on Windows, but at least on Mac OS it does work well with `kqueue(2)`. So, on Linux we are basically stuck making believe that regular files are immediately readable and writable, which is as much as `select` does for us anyway. The futures do work properly on pipes like stdin. On Windows, tokio doesn't give us anything like AsyncFd, so we fall back on using the cap-std-sync windows poll_oneoff inside a block_in_place.

`wasmtime-wasi` now exports two different `Wasi`s - one under `sync::` when the `sync` feature is enabled (which it is by default), and one under `tokio::` when the `tokio` feature is enabled (not a default). The sync implementation is re-exported at the root of the crate, so that existing users do not have to change their code.

A new top-level `examples/tokio` shows using `wasmtime_wasi::tokio::Wasi` under `tokio`. It incorporates a lot of comments @alexcrichton wrote previously.

### CODE REVIEW

File: crates/misc/run-examples/src/main.rs [hunk 1]
```diff
@@ -45,10 +49,17 @@ fn main() -> anyhow::Result<()> {
...
         println!("======== Rust example `{}` ============", example);
-        run(Command::new("cargo")
-            .arg("run")
-            .arg("--example")
-            .arg(&example))?;
+        let mut cargo_cmd = Command::new("cargo");
+        cargo_cmd.arg("run").arg("--example").arg(&example);
+        if is_dir {
+            let mut features_path = std::path::PathBuf::from(example_path);
+            features_path.push("CARGO_FEATURES");
```
Comment (alexcrichton) [inline crates/misc/run-examples/src/main.rs:56]: We probably don't need to be so general here, could this detect maybe "async" in the name and pass a hardcoded feature?

(mostly trying to keep the examples dir relatively clean)

File: crates/test-programs/tests/wasm_tests/runtime/tokio.rs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    let stdout_ = stdout.clone();
+    let stderr = WritePipe::new_in_memory();
+    let stderr_ = stderr.clone();
+
+    let r = tokio::runtime::Runtime::new()
+        .expect("create runtime")
+        .block_on(async move {
+            let mut config = Config::new();
+            config.async_support(true);
+            config.consume_fuel(true);
```
Comment (alexcrichton) [inline crates/test-programs/tests/wasm_tests/runtime/tokio.rs:18]: This isn't needed for tests though, right?

File: crates/test-programs/tests/wasm_tests/runtime/tokio.rs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+            #[cfg(all(unix, not(target_os = "macos")))]
+            {
+                builder = builder.env("ERRNO_MODE_UNIX", "1")?;
+            }
+            #[cfg(target_os = "macos")]
+            {
+                builder = builder
+                    .env("ERRNO_MODE_MACOS", "1")?
+                    .env("NO_FD_ALLOCATE", "1")?;
+            }
```
Comment (alexcrichton) [inline crates/test-programs/tests/wasm_tests/runtime/tokio.rs:56]: This is my personal first time seeing this, although I see that this is simply copied over from the cap-std-sync version. Could this be deduplicated between those two versions perhaps? (also if you know what this does can you add a comment here about that? I have no idea what this is configuring...)

File: crates/test-programs/tests/wasm_tests/runtime/tokio.rs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+            store.out_of_fuel_async_yield(u32::MAX, 10000);
+            Wasi::set_context(&store, builder.build()?)
+                .map_err(|_| anyhow::anyhow!("wasi set_context failed"))?;
+
+            let module =
+                Module::new(store.engine(), &data).context("failed to create wasm module")?;
+            let linker = Linker::new(&store);
+            let instance = linker.instantiate_async(&module).await?;
+            let start = instance.get_typed_func::<(), ()>("_start")?;
+            start.call_async(()).await.map_err(anyhow::Error::from)
```
Comment (alexcrichton) [inline crates/test-programs/tests/wasm_tests/runtime/tokio.rs:131]: Similar to above it seems like much of this could probably be shared between at least the above function, if not with the cap-std-sync verison too ideally

File: crates/wasi-common/cap-std-sync/Cargo.toml [hunk 5]
```diff
@@ -13,13 +13,14 @@ include = ["src/**/*", "LICENSE" ]
 
 [dependencies]
 wasi-common = { path = "../", version = "0.26.0" }
+wiggle = { path = "../../wiggle", version = "0.26.0" }
```
Comment (alexcrichton) [inline crates/wasi-common/cap-std-sync/Cargo.toml:16]: I think this is only used for the `async_trait` macro, right? If that's the case could this be just the `async_trait` dependency?

File: crates/wasi-common/tokio/src/sched/unix.rs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+impl<'a, T> Future for FirstReady<'a, T> {
+    type Output = T;
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> FPoll<T> {
+        let mut result = FPoll::Pending;
+        for f in self.as_mut().0.iter_mut() {
+            match f.as_mut().poll(cx) {
+                FPoll::Ready(r) => match result {
+                    // First ready gets to set the result. But, continue the loop so all futures
+                    // get the opportunity to become ready.
```
Comment (alexcrichton) [inline crates/wasi-common/tokio/src/sched/unix.rs:31]: How come this continues polling all futures rather than just returning the first ready value immediately?

File: crates/wasi-common/tokio/src/sched/windows.rs [hunk 7]
```diff
@@ -0,0 +1,248 @@
+use crate::asyncify;
```
Comment (alexcrichton) [inline crates/wasi-common/tokio/src/sched/windows.rs:1]: This seems like a pretty intense file that has a helper thread to work with stdio on Windows, could this get commented with a high level overview of how all this works?

(I'm not actually sure off the top of my head why this is necessary)

File: crates/wasi-common/tokio/src/sched/windows.rs [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+    file::WasiFile,
+    sched::{
+        subscription::{RwEventFlags, Subscription},
+        Poll,
+    },
+    Error, ErrorExt,
+};
+
+pub async fn poll_oneoff<'a>(poll: &mut Poll<'a>) -> Result<(), Error> {
+    asyncify(move || poll_oneoff_(poll))
```
Comment (alexcrichton) [inline crates/wasi-common/tokio/src/sched/windows.rs:19]: I'm a little confused by this, how come there's an `asyncify` of an `async` function?

Although reading the above `asyncify` again I kind of understand, but not really.

File: crates/wasi-common/tokio/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+// but always returns Poll::Ready, and executes it in a dummy executor
+// on a blocking thread in the tokio thread pool.
+pub(crate) fn asyncify<'a, F, Fut, T>(f: F) -> Result<T, Error>
+where
+    F: FnOnce() -> Fut + Send + 'a,
+    Fut: Future<Output = Result<T, Error>>,
+    T: Send + 'static,
+{
+    tokio::task::block_in_place(move || unsafe { wiggle::run_in_dummy_executor(f()) })
+}
```
Comment (alexcrichton) [inline crates/wasi-common/tokio/src/lib.rs:109]: I think I'm getting a bit lost in my head thinking about this. This feels pretty backwards where we're "asyncifying" already async code, and I'm not sure that I have enough context for untangling why the backwards feeling is actually ok here.

The idea here is that wasi-common requires everything to be async, but then the "sync" versions just do all the blocking work in their async impls. Then when tokio builds on top of the sync versions again it just rewraps everything in `block_in_place`? Basically the `async` in cap-std-sync is a lie?

I think this may want a more clear comment somewhere if possible?

File: crates/wasi-common/tokio/src/sched/windows.rs [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+            Ok(ready_bytes) => {
+                r.complete(ready_bytes, RwEventFlags::empty());
+                ready = true;
+            }
+            Err(e) => {
+                r.error(e);
+                ready = true;
+            }
+        }
+    }
```
Comment (alexcrichton) [inline crates/wasi-common/tokio/src/sched/windows.rs:108]: I'm sort of confused how this works where the unix version does a bunch of fancy "first ready" stuff but on Windows we skip all that?

File: crates/wiggle/src/lib.rs [hunk 11]
```diff
@@ -953,3 +953,40 @@ impl From<GuestError> for Trap {
         Trap::String(err.to_string())
     }
 }
+
+pub unsafe fn run_in_dummy_executor<F: std::future::Future>(future: F) -> F::Output {
```
Comment (alexcrichton) [inline crates/wiggle/src/lib.rs:957]: I think this technically isn't `unsafe`, right?

File: examples/tokio/main.rs [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+impl Inputs {
+    fn new(env: Environment, name: &str) -> Self {
+        Self {
+            env,
+            name: name.to_owned(),
+        }
+    }
+}
+
+fn run_wasm(inputs: Inputs) -> impl Future<Output = Result<(), Error>> {
```
Comment (alexcrichton) [inline examples/tokio/main.rs:79]: Could this function be shortened to basically "FIXME needs https://github.com/alexcrichton/rfcs-2/blob/new-api/accepted/new-api.md to be safe" or something along those lines?

(probably not too worth maintaining docs here about the unsafety)

File: examples/tokio/main.rs [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+    .map_err(|_| anyhow!("setting wasi context"))?;
+
+    let linker = Linker::new(&store);
+
+    // Instantiate
+    let instance = linker.instantiate_async(&inputs.env.module).await?;
+    instance
+        .get_export("_start")
+        .ok_or_else(|| anyhow!("wasm is a wasi command with export _start"))?
+        .into_func()
```
Comment (alexcrichton) [inline examples/tokio/main.rs:156]: This can all be folded into `get_typed_func` I think?

**Review by alexcrichton — COMMENTED** [review]
👍 Nice!

I think there could be some more comments in a few places, and I've only somewhat lightly reviewed all the tokio scheduler bits but I think this looks pretty good. I keep finding it a bit backwards that all the sync stuff *looks* async but isn't, but it's all at least encapsulated.

File: crates/misc/run-examples/src/main.rs (hunk 1 above)
Comment (pchickey) [reply crates/misc/run-examples/src/main.rs:56]: Sure, I could do that. I figured that putting the features in a file would be more discoverable for someone trying to run it standalone than detecting a part of the name.

File: crates/wasi-common/cap-std-sync/Cargo.toml (hunk 5 above)
Comment (pchickey) [reply crates/wasi-common/cap-std-sync/Cargo.toml:16]: Yep thats the only use. I arbitrarily decided to import it via wiggle everywhere so we wouldn't have to keep versions synced up, but realistically I don't expect async_trait to change meaningfully at this point so I could change to importing it directly.

File: crates/wasi-common/tokio/src/sched/unix.rs (hunk 6 above)
Comment (pchickey) [reply crates/wasi-common/tokio/src/sched/unix.rs:31]: In the corner case where several of your subscriptions are ready immediately, the tests expect all of them to be reported rather than just the first. Reporting is part of the future being polled on here, so I want all of them to get a chance to run before returning. I will improve the comment.

File: crates/wasi-common/tokio/src/sched/windows.rs (hunk 7 above)
Comment (pchickey) [reply crates/wasi-common/tokio/src/sched/windows.rs:1]: This code is borrowed from cap-std-sync which is in turn borrowed from the old version of wasi-common. I also don't quite understand it, and this code was highly resistant to some earlier attempts to improve it, so I have just been dragging it along. I can add better comments, though.

File: crates/wasi-common/tokio/src/lib.rs (hunk 9 above)
Comment (pchickey) [reply crates/wasi-common/tokio/src/lib.rs:109]: Agreed this needs a comment. The `async` in cap-std-sync is indeed a lie.

File: crates/wasi-common/tokio/src/sched/windows.rs (hunk 8 above)
Comment (pchickey) [reply crates/wasi-common/tokio/src/sched/windows.rs:19]: `block_on_dummy_executor`?

File: crates/wiggle/src/lib.rs (hunk 11 above)
Comment (pchickey) [reply crates/wiggle/src/lib.rs:957]: Probably not - there is some `unsafe` under the hood but afaict that shouldnt make it unsafe itself. I debated about that keyword and ultimately threw it in there to flag "don't use this unless you are really sure", initially this was totally private code but I needed to reuse it in several crates.

File: crates/test-programs/tests/wasm_tests/runtime/tokio.rs (hunk 4 above)
Comment (pchickey) [reply crates/test-programs/tests/wasm_tests/runtime/tokio.rs:131]: this one is async and cap-std-syncs is sync, but yeah they can be shared locally

File: crates/wasi-common/tokio/src/sched/windows.rs (hunk 10 above)
Comment (pchickey) [reply crates/wasi-common/tokio/src/sched/windows.rs]: I ended up refactoring this to re-use the other implementation without copy-paste, filed #2880, and added some comments.

**Review by alexcrichton — APPROVED** [review]
Ok all looks good to me! I think it's ok to leave the windows poll issue to, well, an issue for now. Other than that seems good to go for me! (although there's some CI failures)
